File: terminal/UserInput.pas
unit UserInput;

interface
uses sysUtils;

function ReadString(prompt: String): String;
function ReadDouble(prompt: String): Double;
function ReadDoubleRange(prompt: String; min, max: Double): Double;
function ReadInteger(prompt: String): Integer;
function ReadIntegerRange(prompt: String; min, max: Integer): Integer;

implementation

	function ReadString(prompt: String): String;
	begin
		Write(prompt, ' ');
		ReadLn(result);
	end;

	function ReadDouble(prompt: String): Double;
	var
		userInput: String;
	begin
		userInput := ReadString(prompt);

		while not TryStrToFloat(userInput, result) do
		begin
			WriteLn('Please enter a number: ');
			userInput := ReadString(prompt);
		end;
	end;

	function ReadDoubleRange(prompt: String; min, max: Double): Double;
	begin
		result := ReadDouble(prompt);
		while (result < min) or (result > max) do
		begin
			WriteLn('Please enter a number between ', min, ' and ', max,'.');
			result := ReadDouble(prompt);
		end;
	end;

	function ReadInteger(prompt: String): Integer;
	var
		userInput: String;
	begin
		userInput := ReadString(prompt);

		while not TryStrToInt(userInput, result) do
		begin
			WriteLn('Please enter a whole number: ');
			userInput := ReadString(prompt);
		end;
	end;

	function ReadIntegerRange(prompt: String; min, max: Integer): Integer;
	begin
		result := ReadInteger(prompt);
		while (result < min) or (result > max) do
		begin
			WriteLn('Please enter a number between ', min, ' and ', max,'.');
			result := ReadInteger(prompt);
		end;
	end;
end.

File: terminal/user_input.c
#include <stdio.h>
#include "user_input.h"

input_string read_string(const char* prompt)
{
  input_string result;    
  printf("%s", prompt); 
  scanf(" %255[^\n]", result.str ); 
  return result; 
 } 


int read_integer(const char* prompt)
{
  input_string line;
  int result;    
  char temp; 
  
  line = read_string(prompt);

  while ( sscanf(line.str, " %d %c", &result, &temp) != 1 )
  { 
    printf("Please enter a whole number.\n");
    line = read_string(prompt);
  }
  
  return result;
}

int read_integer_range(const char* prompt, int min, int max)
{
  int result;
  result = read_integer(prompt);
  while (result < min || result > max)
  {
    printf("Please enter a value between %d and %d.\n", min, max);
    result = read_integer(prompt);
  }
  return result;
}

float read_double(const char* prompt)
{
  input_string line;
  float result;
  char temp;

  line = read_string(prompt);

  while(sscanf(line.str, "%f %c", &result, &temp) !=1)
  {
    printf("Please enter a whole number.\n");
    line = read_string(prompt);
  }
  return result;
}

float read_double_range(const char* prompt, double min, double max)
{
  double result;
  result = read_double(prompt);
  while (result < min || result > max)
  {
    printf("Please enter a value between %f and %f.\n", min, max);
    result = read_double(prompt);
  }
  return result;
}